File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
import math

def solution(numbers):
    answer = 0
    
    number_list = []
    
    for i in range(1, len(numbers) + 1):
        permuts = permutations(numbers, i)
        
        # ('1','0','1') -> 101
        for p in list(permuts):
            string = "".join(p)
            number_list.append(int(string))
        
    number_set = set(number_list)
    arr = list(number_set)
    print(arr)
    
    # 소수인지 판별
    for a in arr:
        # 0, 1 은 소수 아님
        if a <= 1:
            continue
            
        flag = True
        for i in range(2, int(math.sqrt(a)) + 1):
            if a % i == 0:
                flag = False
                break
        # 소수라면 answer 증가
        if flag:
            print("소수임둥", a)
            answer += 1
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.Set;
import java.util.HashSet;


class Solution {
    
    /*
    1. 숫자를 조합(문자열 -> int)
    2. 소수 판별
    3. 정답 저장(HashSet)
    */
    
    public boolean[] isPrime;
    
    public void isPrime_fun(int n){ 
        isPrime = new boolean[n+1];
        
        for(int i = 0; i < isPrime.length; i++){
            isPrime[i] = true;
        }
        
        isPrime[0] = isPrime[1] = false; 
        
        for(int i = 2; i <= Math.sqrt(n); i++){ 
            if(isPrime[i]){
                // i의 배수는 소수가 아니다
                for(int j = i*i; j <= n; j += i){
                    isPrime[j] = false;
                }
            }
        }
    } 
    
    
    public Set<Integer> combis = new HashSet<>();
    
    // n개를 골라 조합을 만들어주는 거
    public void combination(int depth, boolean[] visited, String str, int n, String input){
        if(depth == n) {
            combis.add(Integer.parseInt(str));
            return;
        }
        
        for (int i = 0; i < input.length(); i++){
            if (visited[i]){
                continue;
            }
            visited[i] = true;
            combination(depth + 1, visited, str + input.charAt(i), n, input);
            visited[i] = false;
        }
    } 
    
    
    public int solution(String numbers) {
        int answer = 0;
        int n = numbers.length();
        
        // 1. 숫자를 조합(문자열 -> int)        
        boolean[] visited = new boolean[n];
        
        for (int i=1;i <= n; i++){
            combination(0, visited, "", i, numbers);
        }
        
        // 2. 소수 판별
        isPrime_fun(9999999);
        
        Integer[] arr = combis.toArray(new Integer[0]);
        
        for (int i=0;i < arr.length;i++){
            if(isPrime[arr[i]]){
                answer++;
            }
        }        
        
        return answer;
    }
}